include/revng/Model/DefinedType.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Type.h"
+
+/* TUPLE-TREE-YAML
+name: DefinedType
+type: struct
+inherits: Type
+fields:
+  - name: Definition
+    reference:
+      pointeeType: TypeDefinition
+      rootType: Binary
+TUPLE-TREE-YAML */
+
+#include "revng/Model/Generated/Early/DefinedType.h"
+
+class model::DefinedType : public model::generated::DefinedType {
+public:
+  using generated::DefinedType::DefinedType;
+
+  static UpcastableType make(TypeOfDefinition &&Definition) {
```

Comment:
Requiring a rvalue reference is a bit too strict. Usually we take a copy and let the compiler use the move constructor it on the call site you have a temporary.

Reply:
IMO, I pretty much see this `make` as a pretty internal interface. In the general case, when pushing a type into a binary, user should already get the result of this "make" call instead of the raw reference this accepts.

Reply:
Creating a `QualifiedType` is rather common, I wouldn't say it's low level. Look for all the calls to its constructor, this interface basically replaces that.

But overall, this looks like early optimization, taking an instance and letting the compiler choose if the move constructor is possible should take home 99% of the result.

Reply:
I guess. Either way, let's see how replacement goes first. It's not like removing `&&` from the prototype is going to break the callees.